File: PokemonApp.Application/Services/PokemonService.cs
﻿using PokemonApp.Application.Pokemons;
using PokemonApp.Domain.Entities;
using PokemonApp.Infrastructure.Repositories;

namespace PokemonApp.API.Services
{
    public class PokemonService : IPokemonService
    {
        private readonly IPokemonRepository _pokemonRepository;

        public PokemonService(IPokemonRepository pokemonRepository)
        {
            _pokemonRepository = pokemonRepository;
        }

        public Pokemon GetPokemonById(int id)
        {
            return _pokemonRepository.GetPokemonById(id);
        }

        public Pokemon GetPokemonByName(string name)
        {
            return _pokemonRepository.GetPokemonByName(name);
        }

        public IEnumerable<PokemonDto> GetPokemonsForDashboard()
        {
            return _pokemonRepository.GetPokemonsForDashboard();
        }

        public PokemonSummaryData GetPokemonSummary()
        {
            var pokemons = _pokemonRepository.GetPokemonsForDashboard();

            var totalSpecies = pokemons.Count();

            var typesCounts = pokemons
                .SelectMany(p => p.Types)
                .GroupBy(t => t)
                .Select(g => new { Type = g.Key, Count = g.Count() })
                .ToDictionary(g => g.Type, g => g.Count);

            var generationCounts = pokemons
                .GroupBy(p => p.Generation)
                .Select(g => new { Generation = g.Key, Count = g.Count() })
                .ToDictionary(g => g.Generation, g => g.Count);

            return new PokemonSummaryData
            {
                TotalSpecies = totalSpecies,
                TypeCounts = typesCounts,
                GenerationCounts = generationCounts
            };
        }
    }
}

File: client-app/src/services/PokemonService.ts
import axios from 'axios';
import { Pokemon } from '../types/pokemon';
import { PokemonSummaryData } from '../types/summary';

const API_URL = 'https://localhost:7123/pokemon/';

export const getPokemonSummary = async () => {
  const response = await axios.get<PokemonSummaryData>(`${API_URL}summary`);
  return response.data;
};

export const getPokemonList = async () => {
  const response = await axios.get<Pokemon[]>(`${API_URL}`);
  return response.data;
};

export const getPokemonDetails = async (id: number) => {
  const response = await axios.get<Pokemon>(`${API_URL}${id}`);
  return response.data;
};

export const getPokemonByName = async (name: string) => {
  const response = await axios.get<Pokemon>(`${API_URL}name/${name}`);
  return response.data;
};